lib/ecto/association.ex

```diff
@@ -176,97 +175,163 @@ defmodule Ecto.Association do
   end
 
   @doc """
-  Build an association query through with starting the given reflection
-  and through the given associations.
+  Build an association query through the given associations from the specified owner table
+  and through the given associations. Finally filter by the provided values of the owner_key of
+  the first relationship in the chain. Used in Ecto.assoc/2.
   """
-  def assoc_query(refl, through, query, values)
-
-  def assoc_query(%{owner: owner, through: [h|t], field: field}, extra, query, values) do
-    refl = owner.__schema__(:association, h) ||
-            raise "unknown association `#{h}` for `#{inspect owner}` (used by through association `#{field}`)"
-    assoc_query refl, t ++ extra, query, values
+  def filter_through_chain(owner, through, values) do
+    chain_through(owner, through, nil, values)
+    |> distinct([x], true)
   end
 
-  def assoc_query(%module{} = refl, [], query, values) do
-    module.assoc_query(refl, query, values)
-  end
+  @doc """
+  Join the target table given a list of associations to go through starting from the owner table.
+  """
+  def join_through_chain(owner, through, query) do
+    chain_through(owner, through, query, nil)
+  end
+
+  # This function is used by both join_through_chain/3 and filter_through_chain/3 since the algorithm for both
+  # is nearly identical barring a few differences.
+  defp chain_through(owner, through, join_to, values) do
+    # Flatten the chain of throughs. If any of the associations is a HasThrough this allows us to expand it so we have
+    # a list of atomic associations to join through.
+    {_, through} = flatten_through_chain(owner, through, [])
+
+    # If we're joining then we're going forward from the owner table to the destination table.
+    # Otherwise we're going backward from the destination table then filtering by values.
+    chain_direction = if(join_to != nil, do: :forward, else: :backward)
+
+    # This stage produces a list of joins represented as a keyword list with the following structure:
+    # [
+    #   [schema: (The Schema), in_key: (The key used to join into the table), out_key: (The key used to join with the next), where: (The condition KW list)]
+    # ]
+    relation_list = resolve_through_tables(owner, through, chain_direction)
+
+    # Filter out the joins which are redundant
+    filtered_list = Enum.with_index(relation_list)
+    |> Enum.filter(fn
+      # We always keep the first table in the chain since it's our source table for the query
+      {_, 0} -> true
+
+      {rel, _} ->
+        condition = Keyword.get(rel, :where, [])
+        in_key = Keyword.get(rel, :in_key)
+        out_key = Keyword.get(rel, :out_key)
+
+        # If the condition is not empty we need to join to the table. Otherwise if the in_key and out_key is the same
+        # then this join is redundant.
+        in_key != out_key || condition != []
```

Comment:
Since both sides are booleans:

```suggestion
        in_key != out_key or condition != []
```